phonebook-api/src/main/java/com/telran/phonebookapi/controller/PhoneController.java

```diff
@@ -1,45 +1,83 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.userdetails.UserDetails;
 import org.springframework.web.bind.annotation.*;
 
+import javax.persistence.EntityNotFoundException;
 import javax.validation.Valid;
 import java.util.List;
+import java.util.stream.Collectors;
 
 @RestController
-@CrossOrigin
 @RequestMapping("/api/phone")
 public class PhoneController {
 
+    static final String INVALID_ACCESS = "Error! You have no permission";
+
     PhoneService phoneService;
+    ContactService contactService;
+    PhoneMapper phoneMapper;
 
-    public PhoneController(PhoneService phoneService) {
+    public PhoneController(PhoneService phoneService, ContactService contactService, PhoneMapper phoneMapper) {
         this.phoneService = phoneService;
+        this.contactService = contactService;
+        this.phoneMapper = phoneMapper;
     }
 
     @PostMapping("")
-    public void addPhone(@RequestBody @Valid PhoneDto phoneDto) {
-        phoneService.add(phoneDto);
+    public void addPhone(Authentication auth, @RequestBody @Valid PhoneDto phoneDto) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        Contact contact = contactService.getById(phoneDto.contactId);
+        if (!contact.getUser().getEmail().equals(userDetails.getUsername())) {
+            throw new EntityNotFoundException(INVALID_ACCESS);
+        }
+        phoneService.add(phoneDto.contactId, phoneDto.countryCode, phoneDto.phoneNumber);
     }
 
     @PutMapping("")
-    public void editPhone(@RequestBody @Valid PhoneDto phoneDto) {
-        phoneService.editAllFields(phoneDto);
+    public void editPhone(Authentication auth, @RequestBody @Valid PhoneDto phoneDto) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        Contact contact = contactService.getById(phoneDto.contactId);
+        if (!contact.getUser().getEmail().equals(userDetails.getUsername())) {
+            throw new EntityNotFoundException(INVALID_ACCESS);
+        }
+        phoneService.editAllFields(phoneDto.contactId, phoneDto.countryCode, phoneDto.phoneNumber);
     }
 
     @GetMapping("/{id}")
-    public PhoneDto getById(@PathVariable int id) {
-        return phoneService.getById(id);
+    public PhoneDto getById(Authentication auth, @PathVariable int id) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        Contact contact = contactService.getById(id);
+        if (!contact.getUser().getEmail().equals(userDetails.getUsername())) {
+            throw new EntityNotFoundException(INVALID_ACCESS);
+        }
+        return phoneMapper.mapPhoneToDto(phoneService.getById(id));
     }
 
     @DeleteMapping("/{id}")
-    public void removeById(@PathVariable int id) {
+    public void removeById(Authentication auth, @PathVariable int id) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        Contact contact = contactService.getById(id);
```

Comment:
wrong. the id is not an id of a contact, it is id of a phone. Check everywhere

Reply:
resolved